src/adc.rs

```diff
@@ -0,0 +1,562 @@
+/// TODO: Extend/generalize beyond f303
+macro_rules! adc_hal {
+    ($(
+            $ADC:ident: ($adcx:ident, $ADC_COMMON:ident),
+    )+) => {
+        $(
+            impl Adc<$ADC> {
+
+                /// Init a new ADC
+                ///
+                /// Enables the clock, performs a calibration and enables the ADC
+                pub fn $adcx(
+                    rb: $ADC,
+                    adc_common : &mut $ADC_COMMON,
+                    ahb: &mut AHB,
+                    ckmode: CKMODE,
+                    clocks: Clocks,
+                ) -> Option<Self> {
+                    let mut this_adc = Self {
+                        rb,
+                        clocks,
+                        ckmode,
+                        operation_mode: None,
+                    };
+                    if !(this_adc.clocks_welldefined(clocks)) {
+                        return None;
+                    }
+                    if !(this_adc.enable_clock(ahb, adc_common)){
+                        return None;
+                    }
+                    this_adc.set_align(Align::default());
+                    this_adc.calibrate();
+                    // ADEN bit cannot be set during ADCAL=1
+                    // and 4 ADC clock cycle after the ADCAL
+                    // bit is cleared by hardware
+                    this_adc.wait_adc_clk_cycles(4);
+                    this_adc.enable();
+                    Some(this_adc)
+                }
+
+                /// Software can use CKMODE::SYNCDIV1 only if
+                /// hclk and sysclk are the same. (see reference manual 15.3.3)
+                fn clocks_welldefined(&self, clocks: Clocks) -> bool {
+                    !(self.ckmode == CKMODE::SYNCDIV1 && !(clocks.hclk().0 == clocks.sysclk().0))
+                }
+
+                /// sets up adc in one shot mode for a single channel
+                pub fn setup_oneshot(&mut self) {
+                    // stop and clear overrun events
+                    self.rb.cr.modify(|_, w| w.adstp().set_bit());
+                    self.rb.isr.modify(|_, w| w.ovr().clear_bit());
+
+                    self.rb.cfgr.modify(|_, w| w
+                        .cont().clear_bit()
+                        .ovrmod().clear_bit()
+                    );
+
+                    self.rb.sqr1.modify(|_, w| w.l().bits(0));
+
+                    self.operation_mode = Some(OperationMode::OneShot);
+                }
+
+                fn set_align(&self, align: Align) {
+                    self.rb.cfgr.modify(|_, w| w.align().bit(align.bitvalue()));
+                }
+
+                fn enable(&mut self) {
+                    self.rb.cr.modify(|_, w| w.aden().set_bit());
+                    while self.rb.isr.read().adrdy().bit_is_clear() {}
+                }
+
+                fn disable(&mut self) {
+                    self.rb.cr.modify(|_, w| w.aden().clear_bit());
+                }
+
+
```

Comment:
Superfluous newline. Pretty sure the only reason rustfmt is okay with this is because it doesn't handle macros